stream-projects/notifications/functions/src/notify-is-live.ts

```diff
@@ -0,0 +1,139 @@
+import { serve } from "https://deno.land/std@0.159.0/http/server.ts";
+import { fetchChannel } from "./util/channel.ts";
+import { fetchKV, upsertKV } from "./util/kv.ts";
+import {
+  NotificationJobUpsertInput,
+  upsertNotificationJob,
+} from "./util/notification.ts";
+
+const HAS_SENT_NOTIFICATION_KEY = "hasSentNotificationSinceLive";
+
+function makeResp(statusCode: number, body: unknown) {
+  return new Response(JSON.stringify(body), {
+    status: statusCode,
+    headers: {
+      "Content-Type": "application/json",
+    },
+  });
+}
+
+interface CheckIsLiveReqBody {
+  runtimeData?: {
+    notificationJobUpsertInput?: NotificationJobUpsertInput;
+  };
+}
+
+serve(
+  async (req) => {
+    const accessToken = req.headers.get("x-access-token");
+    const orgId = req.headers.get("x-org-id");
+    let reqBody: CheckIsLiveReqBody;
+
+    try {
+      reqBody = await req.json();
+    } catch {
+      return makeResp(400, {
+        error: true,
+        message: "Please provide a valid JSON body.",
+      });
+    }
+
+    // return an error if they don't provide an access token
+    if (!accessToken) {
+      return makeResp(400, {
+        error: true,
+        message:
+          "Please provide a user access token via the 'x-access-token' header.",
+      });
+    }
+
+    // return an error if they don't provide an access token
+    if (!orgId) {
+      return makeResp(400, {
+        error: true,
+        message: "Please provide an org id via the 'x-org-id' header.",
+      });
+    }
+
+    // grab the notificationJobUpsertInput from the body
+    const { notificationJobUpsertInput } = reqBody?.runtimeData ?? {};
+    if (!notificationJobUpsertInput) {
+      return makeResp(400, {
+        error: true,
+        message: "Please provide a notificationJobUpsertInput object.",
+      });
+    }
+
+    // fetch the associated channel from mycelium
+    const channel = await fetchChannel(accessToken, orgId);
+    if (!channel) {
+      return makeResp(404, {
+        error: true,
+        message: "The channel requested from the org is undefined.",
+      });
+    }
+
+    // fetch `hasSentNotificationSinceLive` KV so we know if we need to send a notification
+    const hasSentKV = await fetchKV(
+      HAS_SENT_NOTIFICATION_KEY,
```

Comment:
does this need to be generated based on the current stream videoId or how does this reset from stream to stream?

Reply:
If the function is triggered, and the stream is not live, then it resets the flag to false. The next time the stream goes live, it sees that the flag is false, sets it to true, and sends notifications to users. Then every subsequent time the function is triggered during the stream, it sees that the channel is live, but it also sees that the flag is true, so it doesn't send a notification to users.

By flag I mean `hasSentNotificationSinceLive`.